### PR TITLE
Pkcs11 fixes 2

### PR DESCRIPTION
**Pull request by SparkiDev**

### CODE REVIEW

**Comment by toddouska** [comment]
@dgarske please review, after approval assign to me.

File: src/internal.c [hunk 1]
```diff
@@ -20047,6 +20052,76 @@ int PickHashSigAlgo(WOLFSSL* ssl, const byte* hashSigAlgo, word32 hashSigAlgoSz)
...
+int CreateDevPrivateKey(void** pkey, byte* data, word32 length, int hsType,
+                        int label, int id, void* heap, int devId)
+{
+    int ret = NOT_COMPILED_IN;
+
+    if (hsType == DYNAMIC_TYPE_RSA) {
+#ifndef NO_RSA
+        RsaKey* rsaKey;
+
+        rsaKey = XMALLOC(sizeof(RsaKey), heap, DYNAMIC_TYPE_RSA);
```
Comment (dgarske) [inline src/internal.c:20078]: This will need a cast on the XMALLOC for some compilers.

File: src/internal.c [hunk 2]
```diff
@@ -20047,6 +20052,76 @@ int PickHashSigAlgo(WOLFSSL* ssl, const byte* hashSigAlgo, word32 hashSigAlgoSz)
...
+        else {
+            XFREE(rsaKey, heap, DYNAMIC_TYPE_RSA);
+        }
+#endif
+    }
+    else if (hsType == DYNAMIC_TYPE_ECC) {
+#ifdef HAVE_ECC
+        ecc_key* ecKey;
+
+        ecKey = XMALLOC(sizeof(ecc_key), heap, DYNAMIC_TYPE_ECC);
```
Comment (dgarske) [inline src/internal.c:20101]: Same here with cast on XMALLOC.

File: src/ssl.c [hunk 3]
```diff
@@ -7377,7 +7353,41 @@ int wolfSSL_CTX_check_private_key(const WOLFSSL_CTX* ctx)
...
+            ret = wc_CryptoCb_RsaCheckPrivKey(pkey, der->publicKey,
+                                              der->pubKeySize);
+            if (ret == 0)
+                ret = 1;
+            wc_FreeRsaKey(pkey);
+        }
+        else if (ret == 0 && der->keyOID == ECDSAk) {
+            ret = wc_CryptoCb_EccCheckPrivKey(pkey, der->publicKey,
+                                              der->pubKeySize);
+            if (ret == 0)
```
Comment (dgarske) [inline src/ssl.c:7380]: This logic for taking return code 0 and converting to 1 then checking for "1" and making it `WOLFSSL_SUCCESS` seems clunky. Is it possible to improve this logic? Same below in `wolfSSL_check_private_key`.

File: wolfcrypt/src/wc_pkcs11.c [hunk 4]
```diff
@@ -747,9 +784,10 @@ static int Pkcs11CreateRsaPrivateKey(CK_OBJECT_HANDLE* privateKey,
         { CKA_EXPONENT_2,       NULL,          0                    },
         { CKA_COEFFICIENT,      NULL,          0                    },
         { CKA_PUBLIC_EXPONENT,  NULL,          0                    },
-        { CKA_ID,               NULL,          0                    }
+        { 0,                    NULL,          0                    },
+        { 0,                    NULL,          0                    }
     };
-    CK_ULONG        keyTmplCnt = sizeof(keyTemplate) / sizeof(*keyTemplate) - 1;
+    CK_ULONG        keyTmplCnt = sizeof(keyTemplate) / sizeof(*keyTemplate) - 2;
```
Comment (dgarske) [inline wolfcrypt/src/wc_pkcs11.c:790]: Can you clarify why the empty entries above are required and the `-2`?

File: wolfcrypt/src/wc_pkcs11.c [hunk 5]
```diff
@@ -843,9 +977,24 @@ static int Pkcs11CreateEccPrivateKey(CK_OBJECT_HANDLE* privateKey,
         { CKA_KEY_TYPE,  &ecKeyType,    sizeof(ecKeyType)    },
         { operation,     &ckTrue,       sizeof(ckTrue)       },
         { CKA_EC_PARAMS, params,        0                    },
-        { CKA_VALUE,     NULL,          0                    }
+        { CKA_VALUE,     NULL,          0                    },
+        { 0,             NULL,          0                    },
+        { 0,             NULL,          0                    }
     };
-    CK_ULONG        keyTmplCnt = sizeof(keyTemplate) / sizeof(*keyTemplate);
+    CK_ULONG        keyTmplCnt = sizeof(keyTemplate) / sizeof(*keyTemplate) - 2;
```
Comment (dgarske) [inline wolfcrypt/src/wc_pkcs11.c:984]: Can you clarify why the empty entries above are required and the `-2`?

File: wolfcrypt/src/wc_pkcs11.c [hunk 6]
```diff
@@ -820,6 +865,95 @@ static int Pkcs11EccSetParams(ecc_key* key, CK_ATTRIBUTE* tmpl, int idx)
...
+    CK_ATTRIBUTE    keyTemplate[] = {
+        { CKA_CLASS,     &pubKeyClass, sizeof(pubKeyClass) },
+        { CKA_KEY_TYPE,  &ecKeyType,   sizeof(ecKeyType)   },
+        { operation,     &ckTrue,      sizeof(ckTrue)      },
+        { CKA_EC_PARAMS, params,       0                   },
+        { CKA_EC_POINT,  NULL,         0                   },
+        { 0,             NULL,         0                   },
+        { 0,             NULL,         0                   }
+    };
+    CK_ULONG        keyTmplCnt = sizeof(keyTemplate) / sizeof(*keyTemplate) - 2;
```
Comment (dgarske) [inline wolfcrypt/src/wc_pkcs11.c:900]: Can you clarify why the empty entries above are required and the `-2`?

**Review by dgarske — CHANGES_REQUESTED** [review]
Hey Sean, just a few minor items. If these changes tie back to a ZD ticket, will you put that in the comments?

File: src/internal.c (hunk 1 above)
Comment (SparkiDev) [reply src/internal.c:20078]: Fixed

File: src/internal.c (hunk 2 above)
Comment (SparkiDev) [reply src/internal.c:20101]: Fixed

File: src/ssl.c (hunk 3 above)
Comment (SparkiDev) [reply src/ssl.c:7380]: Reworked.

File: wolfcrypt/src/wc_pkcs11.c (hunk 4 above)
Comment (SparkiDev) [reply wolfcrypt/src/wc_pkcs11.c:790]: Comments added.

File: wolfcrypt/src/wc_pkcs11.c (hunk 5 above)
Comment (SparkiDev) [reply wolfcrypt/src/wc_pkcs11.c:984]: Comments added - optional entries.

File: wolfcrypt/src/wc_pkcs11.c (hunk 6 above)
Comment (SparkiDev) [reply wolfcrypt/src/wc_pkcs11.c:900]: Comment added - optional entries.

**Review by dgarske — APPROVED** [review]

**Pull request merged**